src/main/flight/imu.c

```diff
@@ -541,18 +567,31 @@ static void imuCalculateEstimatedAttitude(timeUs_t currentTimeUs)
         useMag = true;
     }
 #endif
+
 #if defined(USE_GPS)
-    if (!useMag && sensors(SENSOR_GPS) && STATE(GPS_FIX) && gpsSol.numSat > GPS_MIN_SAT_COUNT && gpsSol.groundSpeed >= GPS_COG_MIN_GROUNDSPEED) {
-        // Use GPS course over ground to correct attitude.values.yaw
-        const float imuYawGroundspeed = fminf (gpsSol.groundSpeed / GPS_COG_MAX_GROUNDSPEED, 2.0f);
-        courseOverGround = DECIDEGREES_TO_RADIANS(gpsSol.groundCourse);
-        cogYawGain = (FLIGHT_MODE(GPS_RESCUE_MODE)) ? gpsRescueGetImuYawCogGain() : imuYawGroundspeed;
-        // normally update yaw heading with GPS data, but when in a Rescue, modify the IMU yaw gain dynamically
-        if (shouldInitializeGPSHeading()) {
-            // Reset our reference and reinitialize quaternion.
-            // shouldInitializeGPSHeading() returns true only once.
+    if (!useMag && sensors(SENSOR_GPS) && STATE(GPS_FIX) && gpsSol.numSat > GPS_MIN_SAT_COUNT) {
+        static bool gpsHeadingInitialized = false;
+        if (gpsHeadingInitialized) {
+            courseOverGround = DECIDEGREES_TO_RADIANS(gpsSol.groundCourse);
+            // used in imuMahonyAHRSupdate()
+            if (FLIGHT_MODE(GPS_RESCUE_MODE)) {
+                // GPS_Rescue adjusts gpsGroundspeedGain during a rescue in a range 0 - 4.5, depending on GPS Rescue state and groundspeed relative to speed to home.
+                gpsGroundspeedGain = gpsRescueGetImuYawCogGain();
+            } else {
+                // in normal flight, IMU should:
+                // - heavily average GPS heading values at low speed, since they are random, almost
+                // - respond more quickly at higher speeds.
+                // GPS typically returns quite good heading estimates at or above 0.5- 1.0 m/s, quite solid by 2m/s
+                // gpsGroundspeedGain will be 0 at 0.0m/s, rising slowly towards 1.0 at 1.0 m/s, and reaching max of 10.0 at 10m/s
+                const float speedRatio = (float)gpsSol.groundSpeed / GPS_COG_MIN_GROUNDSPEED;
+                gpsGroundspeedGain = speedRatio > 1.0f ? fminf(speedRatio, 10.0f) : speedRatio * speedRatio;
+            }
+        } else if (gpsSol.groundSpeed > GPS_COG_MIN_GROUNDSPEED) {
+            // Reset the reference and reinitialize quaternion factors when GPS groundspeed > 2.0 m/s
             imuComputeQuaternionFromRPY(&qP, attitude.values.roll, attitude.values.pitch, gpsSol.groundCourse);
```

Comment:
This will cause large jump in heading, possibly confusing code that depends on heading. But ramping heading by big gain may be even worse, so keep it for now

Reply:
I noticed this and wondered why it was there.  In early logs there was no change in ez_ef or indeed anything I could see when this re-initialisation took place.  It seemed strange but I just left it there.  I figured it needed to be performed when the quad was moving, but in the previous code it was possible that it would have been performed while the quad was still, or while moving, and that doesn't seem right.

Reply:
Ah yes you are right, it did.  Apologies.
@ledvinap Could I ask you more about this 'reset the reference' idea? 
Currently the quad's IMU heading defaults to North (0). 
If the pilot arms and then takes off and flies due North, the IMU heading, and arrow, are already 'correct' from the moment of takeoff, and there is little or no GPS adjustment required.
But if they arm and fly South, then the IMU heading is initially 180 degrees different from the flightpath South.  It then takes quite a long time, at least several seconds, as a result of the IMU lag and the slow initial speed, for the IMU code to rotate the heading from North to South, to align correctly with the flightpath over ground.
It occurred to me that it would be quicker if we could use this once-only 'reset the reference' function to  force IMU heading to match the GPS heading.  Then when we take off, our heading will be wrong until we get some groundspeed, then jump to the initial flightpath, directly.
I wonder if you could perhaps explain how we might do that, at this point in the code?